common/src/main/java/org/apache/cassandra/sidecar/common/CassandraAdapterDelegate.java

```diff
@@ -145,47 +124,73 @@ private void healthCheckInternal()
             Row oneResult = activeSession.execute("select release_version, partitioner from system.local")
                                          .one();
 
-            String releaseVersion = oneResult.getString("release_version");
-            // update the nodeSettings cache.
             // Note that within the scope of this method, we should keep on using the local releaseVersion
-            nodeSettings = new NodeSettings(releaseVersion, oneResult.getString("partitioner"));
-            // this might swap the adapter out
-            SimpleCassandraVersion newVersion = SimpleCassandraVersion.create(releaseVersion);
-            if (!newVersion.equals(currentVersion))
+            String releaseVersion = oneResult.getString("release_version");
+            NodeSettings newNodeSettings = new NodeSettings(releaseVersion, oneResult.getString("partitioner"));
+            if (!newNodeSettings.equals(nodeSettings))
             {
-                currentVersion = newVersion;
-                adapter = versionProvider.getCassandra(nodeSettings.releaseVersion()).create(cqlSession, jmxClient);
-                logger.info("Cassandra version change detected. New adapter loaded: {}", adapter);
+                // update the nodeSettings cache.
+                SimpleCassandraVersion previousVersion = currentVersion;
+                currentVersion = SimpleCassandraVersion.create(releaseVersion);
+                adapter = versionProvider.cassandra(releaseVersion)
+                                         .create(cqlSessionProvider, jmxClient);
+                nodeSettings = newNodeSettings;
```

Comment:
There is a potential race that set `nodeSettings` to an unexpected value. But since it runs periodically, it should be corrected in the following runs. So.. I think it is acceptable.

Reply:
this shouldn't be an issue because this method is guaranteed to only run by a single thread

Reply:
yep